blocks/serialization.py

```diff
@@ -34,130 +142,54 @@
-        :class:`PersistentNdarrayID` saves any :class:`numpy.ndarray` to a
-        separate NPY file inside of the zip file.
-    use_cpickle : bool
-        This enables the use of C-version of `pickle` (known as ``cPickle``
-        in Python 2). Note that this disables warnings about trying to
-        pickle objects in the ``__main__`` namespace.
-
-    Notes
-    -----
-    The final file is simply a zipped file containing at least one file,
-    `pkl`, which contains the pickled object. It can contain any other
-    number of external objects. Note that the zip files are compatible with
-    NumPy's :func:`numpy.load` function.
-
-    >>> import numpy
-    >>> from blocks.bricks import MLP, Identity
-    >>> from blocks.initialization import Constant
-    >>> mlp = MLP([Identity()], [10, 10], weights_init=Constant(0.),
-    ...           biases_init=Constant(0.))
-    >>> mlp.initialize()
-    >>> with open('model.zip', 'wb') as f:
-    ...     dump(mlp, f)
-    >>> 'mlp-linear_0.W' in numpy.load('model.zip').keys()
-    True
-    >>> 'mlp-linear_0.b' in numpy.load('model.zip').keys()
-    True
-    >>> numpy.load('model.zip')['mlp-linear_0.W'].shape
-    (10, 10)
-    >>> with open('model.zip', 'rb') as f:
-    ...     mlp2 = load(f)
-    >>> mlp2  # doctest: +ELLIPSIS
-    <blocks.bricks.MLP object at ...: name=mlp>
+    \*\*kwargs
+        Keyword arguments to be passed to `pickle.Pickler`.
 
     """
-    with closing(zipfile.ZipFile(file_handler, 'w', zipfile.ZIP_DEFLATED,
-                                 allowZip64=True)) as zip_file:
-        def func(f):
-            if use_cpickle:
-                p = cPickle.Pickler(f, protocol=protocol)
-            else:
-                p = PicklerWithWarning(f, protocol=protocol)
-            p.persistent_id = persistent_id(zip_file)
-            p.dump(obj)
-        pkl_utils.zipadd(func, zip_file, 'pkl')
-
-
-# A thin wrapper around Theano load.
-load = pkl_utils.load
+    if use_cpickle:
+        pickler = cPickle.Pickler
+    else:
+        pickler = _PicklerWithWarning
+    with closing(tarfile.TarFile(fileobj=file_, mode='w')) as tar_file:
+        external_objects = {}
+        def _save_parameters(f):
+            renamer = _Renamer()
+            named_parameters = {renamer(p): p for p in parameters}
+            numpy.savez(f, **{n: p.get_value()
+                              for n, p in named_parameters.items()})
+            for n, p in named_parameters.items():
+                array_ = p.container.storage[0]
+                external_objects[id(array_)] = _mangle_parameter_name(
+                    type(array_), n)
+        if parameters:
+            _taradd(_save_parameters, tar_file, '_parameters')
+        def _save_object(f):
+            p = pickler(f, protocol=protocol, **kwargs)
+            p.persistent_id = _PersistentID(external_objects)
+            p.dump(object_)
+        _taradd(_save_object, tar_file, '_pkl')
```

Comment:
There is one very important thing that we forgot here. Pickling the whole object should be optional, because sometimes it is actually the most painful thing. Can you please add a flag for that?